embedded/tbtree/tbtree.go

```diff
@@ -1758,63 +1758,102 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
+func (n *innerNode) insertAt(kvs []*KV, ts uint64) (nodes []node, depth int, err error) {
+	if n.mutated() {
+		return n.updateOnInsertAt(kvs, ts)
+	}
+
+	newNode := &innerNode{
+		t:       n.t,
+		nodes:   make([]node, len(n.nodes)),
+		_ts:     n._ts,
+		_minOff: n._minOff,
+		mut:     true,
 	}
-	return n.updateOnInsertAt(key, value, ts)
-}
 
-func (n *innerNode) updateOnInsertAt(key []byte, value []byte, ts uint64) (nodes []node, depth int, err error) {
-	insertAt := n.indexOf(key)
+	copy(newNode.nodes, n.nodes)
 
-	c := n.nodes[insertAt]
+	return newNode.updateOnInsertAt(kvs, ts)
+}
 
-	cs, depth, err := c.insertAt(key, value, ts)
-	if err != nil {
-		return nil, 0, err
+func (n *innerNode) updateOnInsertAt(kvs []*KV, ts uint64) (nodes []node, depth int, err error) {
+	// group kvs by child at which they will be inserted
+	kvsPerChild := make(map[int][]*KV)
+
+	for _, kv := range kvs {
+		childIndex := n.indexOf(kv.K)
+		kvsPerChild[childIndex] = append(kvsPerChild[childIndex], kv)
 	}
 
-	n._ts = ts
+	var wg sync.WaitGroup
+	wg.Add(len(kvsPerChild))
 
-	ns := make([]node, len(n.nodes)+len(cs)-1)
+	nodesPerChild := make(map[int][]node)
+	var nodesMutex sync.Mutex
 
-	copy(ns, n.nodes[:insertAt])
-	copy(ns[insertAt:], cs)
-	copy(ns[insertAt+len(cs):], n.nodes[insertAt+1:])
+	for childIndex, childKVs := range kvsPerChild {
+		// insert kvs at every child simultaneously
+		go func(childIndex int, childKVs []*KV) {
+			defer wg.Done()
 
-	n.nodes = ns
+			c := n.nodes[childIndex]
 
-	nodes, err = n.split()
+			cs, cdepth, cerr := c.insertAt(childKVs, ts)
 
-	return nodes, depth + 1, err
-}
+			nodesMutex.Lock()
+			defer nodesMutex.Unlock()
+
+			if err != nil {
+				// if any of its children fail to insert, insertion fails
+				err = cerr
+				return
+			}
 
-func (n *innerNode) copyOnInsertAt(key []byte, value []byte, ts uint64) (nodes []node, depth int, err error) {
-	insertAt := n.indexOf(key)
+			nodesPerChild[childIndex] = cs
+			if cdepth > depth {
+				depth = cdepth
+			}
+		}(childIndex, childKVs)
+	}
 
-	c := n.nodes[insertAt]
+	// wait for all the insertions to be done
+	wg.Wait()
 
-	cs, depth, err := c.insertAt(key, value, ts)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	newNode := &innerNode{
-		t:       n.t,
-		nodes:   make([]node, len(n.nodes)+len(cs)-1),
-		_ts:     ts,
-		mut:     true,
-		_minOff: n._minOff,
+	n._ts = ts
+
+	// count the number of children after insertion
+	nsSize := len(n.nodes)
+
+	for i := range n.nodes {
+		cs, ok := nodesPerChild[i]
+		if ok {
+			nsSize += len(cs) - 1
+		}
+	}
+
+	ns := make([]node, nsSize)
+	nsi := 0
+
+	for i, n := range n.nodes {
+		cs, ok := nodesPerChild[i]
+		if ok {
+			copy(ns[nsi:], cs)
+			nsi += len(cs)
+		} else {
+			ns[nsi] = n
+			nsi++
+		}
 	}
 
-	copy(newNode.nodes, n.nodes[:insertAt])
-	copy(newNode.nodes[insertAt:], cs)
-	copy(newNode.nodes[insertAt+len(cs):], n.nodes[insertAt+1:])
+	n.nodes = ns
 
-	nodes, err = newNode.split()
+	nodes, _ = n.split()
```

Comment:
are we ignoring error here intentionally?

Reply:
inner nodes does not return error during split, but I'll add an special error handling